subsys/bluetooth/controller/hci/hci.c

```diff
@@ -6533,13 +6534,20 @@ static void le_per_adv_sync_report(struct pdu_data *pdu_data,
 	data_len_max = ADV_REPORT_EVT_MAX_LEN -
 		       sizeof(struct bt_hci_evt_le_meta_event) -
 		       sizeof(struct bt_hci_evt_le_per_advertising_report);
+	data_len_total = node_rx->hdr.rx_ftr.aux_data_len;
 
 	evt_buf = buf;
 
-	if ((le_event_mask & BT_EVT_MASK_LE_PER_ADVERTISING_REPORT) && accept) {
+	if ((le_event_mask & BT_EVT_MASK_LE_PER_ADVERTISING_REPORT) && accept &&
+	    ((data_len_total - data_len) < CONFIG_BT_CTLR_SCAN_DATA_LEN_MAX)) {
+
+		data_len = MIN(data_len, (CONFIG_BT_CTLR_SCAN_DATA_LEN_MAX +
+					  data_len - data_len_total));
```

Comment:
Not sure I understand this. 

This may conditionally reduce `data_len` to a lower value, but doesn't that truncate the data processed then? i.e. doesn't this make us lose data?

Reply:
Yes, that is the intended requirement that the totak  AD data length reported over a series of HCI Periodic Advertising Report to host does not exceed a total length and the data_status is reported as incomplete if peer is sending large AD data.

Reply:
But does this not just truncate the data then?

Reply:
yes, if the scan max data length is 191 bytes and peer sends say 240 bytes in a PDU, the HCI Periodic Advertising Report will have 191 bytes of AD data with data status as incomplete. This is a Bluetooth HCI Extended Advertising and Periodic Advertising report Conformance requirement.

Reply:
I am sorry, but this sounds odd to me. 

If the peer sends 240 bytes and the max data length is 191, then shouldn't this be sent to the host in multiple reports, instead of being truncated?

If the 191 bytes refer to the maximum number of octets a host can receive (even over multiple reports), then doesn't it make more sense to just discard it? There host doesn't have any use for a truncated report in my opinion.

Reply:
Zephyr does not truncate GATT notifications, nor will the Zephyr host send truncated adv reports to the application: 
```
		if (data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_INCOMPLETE) {
			/* Controller truncated, no more data will come.
			 * We do not need to keep track of this advertiser.
			 * Discard this report.
			 */
			(void)net_buf_pull_mem(buf, evt->length);
			reset_reassembling_advertiser();
			continue;
		}
```
As incomplete LTV structures are very rarely usable, similar to truncated notifications. 

If this is the expected behavior of a controller, then this is fine, but the description of 0b10 "Incomplete, data truncated, no more to come" in the HCI spec is:

> An Event_Type with a data status field of "incomplete, data truncated" indicates that the Controller attempted to receive an AUX_CHAIN_IND PDU but was not successful.

And then the question is whether truncating data based on the `Scan_Max_Data` should be considered "Controller attempted to receive an AUX_CHAIN_IND PDU but was not successful", which I am unsure about.

Reply:
> Zephyr does not truncate GATT notifications

Though this is different subject, this exposes a limitation in Zephyr Host API that blocks a GATT server from sending truncated notifications to GATT client that would follow up with a GATT Read Characterisitic Value and GATT Read Long Characteristic Value. Statements in the ATT protocol:

>If the attribute value is longer than (ATT_MTU-3) octets, then only the first
(ATT_MTU-3) octets of this attributes value can be sent in a notification.
Note: For a client to get a long attribute, it must use the
ATT_READ_BLOB_REQ PDU (see Section 3.4.4.5) following the receipt of
this notification.

> As incomplete LTV structures are very rarely usable, similar to truncated notifications.

This again not concerning the Controller, AD data consists of multiple LTVs supplied by upper layers and just the single AUX_SYNC_IND PDUs in every chain can have sufficiently valid LTVs for an application's usecase even when following AUX_CHAIN_IND are missing or truncated. I am not sure on what basis Host design has decided to discard valid LTVs in truncated reports, limiting application usecases.

> And then the question is whether truncating data based on the `Scan_Max_Data` should be considered "Controller attempted to receive an AUX_CHAIN_IND PDU but was not successful", which I am unsure about.

`Scan_Max_Data` value is a Bluetooth Test Specification requirement to comply to and an IUT is verified that it does not generate any more HCI LE Periodic Advertising Reports that would exceed total AD data length above `Scan_Max_Data`

The discussion here is more a Bluetooth Test Specification requirement that is not mentioned in the Bluetooth Core Specification. There is no error in the PR implementation of `Scan_Max_Data` restriction in HCI LE Periodic Advertising Reports. If the Core or Test Spec or Tester implementation is changed in the future, the Controller implementation can be changed as necessary then.

Reply:
> Though this is different subject, this exposes a limitation in Zephyr Host API that blocks a GATT server from sending truncated notifications to GATT client that would follow up with a GATT Read Characterisitic Value and GATT Read Long Characteristic Value. Statements in the ATT protocol:

It is not limiting at all. as the application can simply truncate the data itself. The point is that the host stack will not silently truncate any data at any time. 

>  I am not sure on what basis Host design has decided to discard valid LTVs in truncated reports, limiting application usecases.

We decided in the current implementation not to send incomplete adv reports to the application, even if the report may contain valid LTVs. The reasoning for this is irrelevant though, as we are discussing controller -> host here, and not host -> application. 

> The discussion here is more a Bluetooth Test Specification requirement that is not mentioned in the Bluetooth Core Specification. There is no error in the PR implementation of Scan_Max_Data restriction in HCI LE Periodic Advertising Reports. 

This is where I am unsure. Does the test specification require you to truncate the data, and send it over HCI if it is larger than `Scan_Max_Data`? This seems odd to me, as there's not really any good way of notifying host of truncated data based on length, as the description of `incomplete, data truncated, no more to come` does not cover this scenario in my opinion. 

If the test specification require you to truncate the data and send to the host with data status = 0b10, then I won't object. If it doesn't then I am in favor not sending anything at all, or add that scenario to the core spec as an erratum. 

So please let me know if this is truly a test spec requirement, and if not then I think we need an additional set of eyes and opinions on this :)

Reply:
Verified that the requirement for truncation is from the LL test spec, and that this is the correct course of action.